# **4. Fonksiyonlar ve Koşullu Dallar (Branching)**

Bu bölümde Python'da **fonksiyonlar** ve **koşullu dallanma** kavramlarını ele alacağız. Fonksiyonlar, birden fazla işlemi tek bir blok içinde organize ederek kod tekrarını azaltmamıza yardımcı olur. Koşullu dallanma ise, programın farklı durumlar için farklı yollar izlemesini sağlar. Bu konular, programların daha modüler ve esnek olmasını sağlar.


## **4.1 Fonksiyon Yazma**

**Fonksiyonlar**, belirli girişler alarak (parametreler) bir işlem gerçekleştirir ve bir çıktı döndürür. Örneğin, matematikte bir fonksiyon şu şekilde tanımlanabilir:

$$ f(x) = x^2 + 2x + 1 $$

Benzer şekilde, Python'da bir fonksiyon tanımlamak için `def` anahtar kelimesi kullanılır.

``` python
def fonksiyon_adi():
    # fonksiyon gövdesi
    # yapılacak işlemler
```

In [ ]:
def parabol(x):
    return x**2+2*x+1
parabol(2)

9

Fonksiyonlar isimleri yazılıarak çağrılır. Önceki örnekte ``parabol()`` ile fonksiyonu çağırabiliriz.

In [ ]:
def selamla():
    print("Merhaba, Tensör Akademi!")
selamla()

Merhaba, Tensör Akademi!


**return Nedir?**

return ifadesi, bir fonksiyonun çalışmasını sonlandırır ve belirtilen değeri fonksiyonun çağrıldığı yere geri döndürür.


## **4.2 Fonksiyon Argümanları ve Yerel Değişkenler**

Bir fonksiyon, birden fazla **argüman (parametre)** alabilir ve işleyebilir. Fonksiyon içinde tanımlanan değişkenler, **yerel değişkenlerdir (local variables)** ve sadece o fonksiyon içinde geçerlidir.

Fonksiyona birden fazla parametre gönderilebilir.
```python
def fonksiyon_adi(parametre1, parametre2, ...):
    # yapılacak işlemler
    return sonuc
```
Çok değişkenli fonksiyonlar gibi düşünülebilir.

In [ ]:
def paraboloid(x,y):
  return x**2+2*x*y+y**2
print(paraboloid(2,3))

25


In [ ]:
def topla(a, b):
    sonuc = a + b
    return sonuc

# Fonksiyonu çağır ve dönen değeri kullan
toplam = topla(5, 3)
print(toplam)

8


In [ ]:
import math as m #modül kullanarak da fonksiyon tanımlanabilir.
def trigo(a,b,c):
  return m.sin(a)+m.cos(b)+m.tan(c)
trigo(1,2,3)

0.2827776051864763

## **4.3 Varsayılan Argümanlar ve Dokümantasyon**

Fonksiyon yazıldığında parametre kullanıcı tarafından verilir.Örneğin
```python
trigo(np.pi,0,np.pi/2)
```
çağırıldığında verdiğimiz değerleri okur. Bazen parametre ön tanımlı olarak verilebilir.
```python
def selam(isim="Ziyaretçi"):
    print("Merhaba", isim)
```
Bunlara **varsayılan değerler** denir. Bu, fonksiyonun belirli argümanları olmadan da çalışmasını sağlar.

In [ ]:
def selamla(isim='Misafir'):
    return f'Merhaba, {isim}!'

print(selamla())  # Merhaba, Misafir!
print(selamla('İnan'))  # Merhaba, İnan!

Merhaba, Misafir!
Merhaba, 5!


## **4.4 Koşullu Dallanma (if-else)**

Koşullu ifadeler, belirli bir **şartın doğru olup olmadığına** bağlı olarak farklı işlemler yapmamızı sağlar.

In [ ]:
def pozitif_mi(x):
    if x > 0:
        return 'Pozitif'
    elif x == 0:
        return 'Sıfır'
    else:
        return 'Negatif'

print(pozitif_mi(10))  # Pozitif
print(pozitif_mi(0))   # Sıfır
print(pozitif_mi(5))  # Negatif

Pozitif
Sıfır
Pozitif


## **4.5 Fonksiyonları Argüman Olarak Kullanma**

Python'da bir fonksiyon, başka bir fonksiyona **argüman olarak** verilebilir.

In [ ]:
def iki_kat(x):
    return x * 2

def uygulama(f, x):
    return f(x)


print(uygulama(iki_kat, 5))

10


In [ ]:
import math as m
def kare(x):
    return x ** 2

def hipotenus(a, b):
    return m.sqrt(kare(a) + kare(b))

print(hipotenus(3, 4))


5.0


##**4.6. Fonksiyon İçinde Fonksiyon**

Bir fonksiyonun içinde başka bir fonksiyon tanımlayabiliriz.
Bu yapı genellikle kapsülleme (encapsulation) veya özel yardımcı fonksiyon gerektiğinde kullanılır.

Örnek:


In [ ]:
def dis_fonksiyon(x):
    def ic_fonksiyon(y):
        return y ** 2
    return ic_fonksiyon(x) + 10

print(dis_fonksiyon(3))


19


## **4.7 Python ile Denklem Çözme**
Python ile denklem çözümlerini yapabiliriz. Çözüm için sayısal yöntemleri kendimiz tanımlayabileceğimiz gibi ``scipy`` gibi hazır paketleri de kullanabiliriz.

Örnek olarak newton yöntemi ile başlayalım.

Newton yöntemi, bir fonksiyonun kökünü (yani ( f(x) = 0 ) denkleminin çözümünü) **türev tabanlı** olarak bulmak için kullanılan sayısal bir yöntemdir.
Yaklaşım, fonksiyonun **teğet doğrularını** kullanarak köke adım adım yaklaşır.

$$
x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)}
$$


In [ ]:
from math import cos

def f(x):
    return cos(x) - x

def newton_yontemi(f, x0, tol=1e-6):
    h = 1e-5  # türev için küçük fark
    while abs(f(x0)) > tol:
        # Sayısal türev: f'(x) ≈ (f(x+h) - f(x)) / h
        x0 = x0 - f(x0) / ((f(x0 + h) - f(x0)) / h)
        print(x0,abs(f(x0))-tol)
    return x0

kok = newton_yontemi(f, 0.5)
print(f'Kök: {kok}')

0.7552216601303748 0.027101036013243013
0.7391416956241585 9.366471035790569e-05
0.7390851340464236 -9.98608788298604e-07
Kök: 0.7390851340464236


Bu örnekte
$( f(x) = \cos(x) - x )$ denkleminin kökü yaklaşık olarak:

$$
x \approx 0.739085
$$

olarak bulunur. Bu, `x = cos(x)` denkleminin sabit noktasını temsil eder.



Python’da denklemleri çözmek için iki temel yaklaşım vardır:

1. **Sayısal (Numerical) yöntemlerle çözüm** – `sympy` kullanmadan, doğrudan fonksiyon ve döngülerle
2. **Sembolik (Symbolic) çözüm** – `sympy` kütüphanesiyle bilinmeyenli denklemleri çözmek

### 🧮 1. Fonksiyon Tanımlayarak Denklemi Temsil Etme

Bir denklemi önce **fonksiyon olarak** tanımlarız:


In [ ]:
def f(x):
    return x**2 - 4*x + 3


### 🔍 2. Sayısal Yöntemle (Kök Arama)

Kök bulmak için `scipy` kullanabiliriz:

In [ ]:
from scipy.optimize import fsolve

def f(x):
    return x**2 - 4*x + 3

kok = fsolve(f, 0)   # 0 tahminiyle başlar
print(kok)

[1.]


### 🧠 3. Sembolik (cebirsel) Yöntem: `sympy`

``sympy`` kütüphanesi ile bir değişken tanımlayıp cebirsel işlemler yapabiliriz.

In [ ]:
import sympy as sp

x = sp.Symbol('x')#sembolü tanımladık
denklem = x**2 - 4*x + 3

kokler = sp.solve(denklem, x)
print(kokler)

[1, 3]


### ⚙️ 4. Birden Fazla Denklem (Denklem Sistemi)
``sympy`` ile birden çok denklemi çözebiliriz.



In [ ]:
import sympy as sp

x, y = sp.symbols('x y')
denklem1 = sp.Eq(x + y, 5)
denklem2 = sp.Eq(x - y, 1)

sonuc = sp.solve((denklem1, denklem2), (x, y))
print(sonuc)

{x: 3, y: 2}


### 🧩 5. Fonksiyonla Çözümü Genelleştirmek

Kendimiz küçük bir “denklem çözücü fonksiyon” da yazabiliriz:


In [ ]:
import sympy as sp

def denklem_coz(ifade, degisken='x'):
    x = sp.Symbol(degisken)
    cozum = sp.solve(ifade, x)
    return cozum

print(denklem_coz('x**2 - 5*x + 6'))

[2, 3]


## **4.7 Test Fonksiyonları ile Doğrulama**

Fonksiyonların doğru çalıştığını doğrulamak için test fonksiyonları yazabiliriz.

In [ ]:
def kare_al(x):
  return x**2
def test_kare_al():
    assert kare_al(2) == 4
    assert kare_al(3) == 9
    print('Tüm testler geçti!')
test_kare_al()

Tüm testler geçti!


## **4.8 Lambda (Anonim) Fonksiyonlar**

Python’da **lambda fonksiyonları**, tek satırlık kısa fonksiyonlar yazmak için kullanılır.
Bu fonksiyonlar **isimsizdir (anonimdir)** ve genellikle kısa süreli işlemlerde kullanılır.
Yapısı;

```python
lambda parametreler : ifade
```
biçimidedir.Buradaki `ifade`, fonksiyonun **geri döndüreceği değeri** belirtir.
`return` anahtar sözcüğü kullanılmaz.
Basit Örnek:



In [ ]:
kare = lambda x: x ** 2
print(kare(5))

### 🔹 **Birden Fazla Parametre**



In [ ]:
topla = lambda a, b: a + b
print(topla(3, 7))

### 🔹 **Lambda ve map() ile Liste İşlemleri**



In [ ]:
sayilar = [1, 2, 3, 4, 5]
kareler = list(map(lambda x: x**2, sayilar))
print(kareler)


### 🔹 **Lambda ve filter() ile Koşullu Seçim**


In [ ]:
sayilar = [1, 2, 3, 4, 5, 6]
ciftler = list(filter(lambda x: x % 2 == 0, sayilar))
print(ciftler)

### 🔹 **Lambda ve sorted() ile Sıralama**





In [ ]:
veri = [(1, 3), (2, 2), (3, 1)]
sirali = sorted(veri, key=lambda x: x[1])
print(sirali)

[(3, 1), (2, 2), (1, 3)]



### 💡 **Ne Zaman Kullanılır?**

* Tek satırlık, basit işlemler gerektiğinde
* `map()`, `filter()`, `sorted()` gibi fonksiyonlarla kısa ifade tanımlamak için
* Kodun okunabilirliğini artırmak için (gereksiz fonksiyon tanımlamadan)

## ⚖️ **4.9 Lambda ve def Arasındaki Farklar**

Python’da hem `def` hem de `lambda` ifadeleri **fonksiyon tanımlamak** için kullanılır.
Ancak amaçları ve kullanım biçimleri farklıdır.

---





### 🧩 **1. Yapı (Sözdizimi) Farkı**

| Özellik            | `def` ile Fonksiyon           | `lambda` Fonksiyonu                       |
| ------------------ | ----------------------------- | ----------------------------------------- |
| **Tanım Şekli**    | Birden fazla satır içerebilir | Tek satırda yazılır                       |
| **Anahtar Sözcük** | `def`                         | `lambda`                                  |
| **İsim**           | Genellikle isimlidir          | İsimsiz (anonim)                          |
| **return**         | Kullanılır                    | Kullanılmaz (ifade sonucu otomatik döner) |

---

### 🔹 **Örnek 1: def ile**

```python
def kare(x):
    return x ** 2
```

### 🔹 **Örnek 2: lambda ile**

```python
kare = lambda x: x ** 2
```

İkisi de aynı sonucu verir:

```python
print(kare(5))  # 25
```


---

### 🧮 **2. Kapsam ve Karmaşıklık**

* `def`: Birden fazla ifade, koşul veya döngü içerebilir.
* `lambda`: Yalnızca **tek bir ifade** (tek satır) içerebilir.

**Örnek:**

```python
# def ile
def kontrol(x):
    if x > 0:
        return "Pozitif"
    else:
        return "Negatif veya sıfır"

# lambda ile (tek satırda)
kontrol = lambda x: "Pozitif" if x > 0 else "Negatif veya sıfır"
```

---

### 🧠 **3. Nerede Kullanılırlar?**

| Kullanım Durumu                                | Tercih     |
| ---------------------------------------------- | ---------- |
| Tek satırlık basit işlem                       | ✅ `lambda` |
| Birden fazla adım veya karmaşık mantık         | ✅ `def`    |
| Sık kullanılacak fonksiyonlar                  | ✅ `def`    |
| Kısa süreli, bir kez kullanılacak fonksiyonlar | ✅ `lambda` |

---



### 🧩 **4. Örnek: Sıralama İşlemi**

```python
veriler = [(1, 3), (2, 2), (3, 1)]

# lambda ile
sirali = sorted(veriler, key=lambda x: x[1])

# def ile
def siralama_kriteri(x):
    return x[1]
sirali2 = sorted(veriler, key=siralama_kriteri)

print(sirali)
print(sirali2)
```

---

### 🎯 **Özet**

| Kriter            | `def`                | `lambda`          |
| ----------------- | -------------------- | ----------------- |
| Satır sayısı      | Çok satırlı olabilir | Tek satır         |
| return            | Kullanılır           | Gerekmez          |
| İsmi              | Genellikle vardır    | Genellikle yoktur |
| Karmaşık işlem    | Uygun                | Uygun değil       |
| Kısa geçici işlem | Gereksiz             | İdeal ✅           |



### **Alıştırmalar**
1. Ekrana “Merhaba, Tensör Akademi!” yazdıran bir fonksiyon tanımlayın ve çalıştırın.

2. Kullanıcıdan iki sayı alıp bu sayıların ortalamasını döndüren bir fonksiyon yazın.

3. Yarıçapı verilen bir dairenin alanını hesaplayan bir fonksiyon tanımlayın.
   $$
   A = \pi r^2
   $$
   Sonucu `return` ifadesiyle döndürün.

4. Bir isim parametresi alan bir fonksiyon tanımlayın.
   Eğer isim verilmezse “Ziyaretçi” yazsın.
   Örneğin `mesaj_yaz("İnan")` ve `mesaj_yaz()` şeklinde iki farklı çağrı yapın.

5. Bir fonksiyon içinde başka bir fonksiyon tanımlayın.
   İç fonksiyon sayının karesini alsın, dış fonksiyon 10 eklesin.
   Sonucu döndürüp ekrana yazdırın.

6. `lambda` ifadesi kullanarak bir sayının küpünü hesaplayan kısa bir fonksiyon yazın. Farklı değerlerle deneyin.

7. `[1, 2, 3, 4, 5]` listesindeki her elemanın karesini almak için `map()` ve `lambda` kullanın.
   Sonuçları listeye çevirip yazdırın.

8. `[3, 8, 12, 5, 7, 10]` listesindeki 5’ten büyük sayıları bulmak için `filter()` ve `lambda` kullanın.
   Sonuçları ekrana yazdırın.

9. ( f(x) = \cos(x) - x ) fonksiyonunun kökünü yaklaşık olarak bulan bir Newton yöntemi fonksiyonu yazın ve sonucu yazdırın.

10. `def` ile tanımlanan bir toplama fonksiyonuyla aynı işi yapan bir `lambda` fonksiyonu tanımlayın.  Her ikisini de çağırarak sonuçlarını karşılaştırın.
11. Bir sayının faktöriyelini hesaplayan bir fonksiyon yazın.  
Örneğin `faktoriyel(5)` çağrıldığında sonuç `120` olmalı.
12. Verilen bir sayının asal olup olmadığını kontrol eden bir fonksiyon yazın.  
Asal ise “Asal sayı”, değilse “Asal değil” yazsın.
13. İki sayı arasındaki tüm sayıların karelerinin toplamını döndüren bir fonksiyon yazın.  
Örneğin `kare_toplami(1, 4)` sonucu $1^2 + 2^2 + 3^2 + 4^2 = 30$ olmalı.
14. Kullanıcıdan bir sayı alın.  
Bu sayının pozitif, negatif veya sıfır olup olmadığını belirleyen bir fonksiyon tanımlayın.
15. Aşağıdaki fonksiyonu tanımlayın:  
$$ f(x) = x^3 - 2x^2 + x - 3 $$
Fonksiyonun belirli bir $x$ değeri için sonucunu hesaplayıp döndürsün.  
Farklı değerlerle deneyin (örneğin $x = -1, 0, 2 $).

